File: go/httpstate.go
// HTTPState, https://httpstate.com/
// Copyright (C) Alex Morales, 2026
//
// Unless otherwise stated in particular files or directories, this software is free software.
// You can redistribute it and/or modify it under the terms of the GNU Affero
// General Public License as published by the Free Software Foundation, either
// version 3 of the License, or (at your option) any later version.

package httpstate

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/gorilla/websocket"
)

type GetArgs struct {
	Authorization string
}

type GetResult struct {
	Data         string
	ETag         string
	LastModified string
}

type SetArgs struct {
	Authorization string
}

func Get(uuid string, args *GetArgs) (*GetResult, error) {
	url := fmt.Sprintf("https://httpstate.com/%s", uuid)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Println(time.Now().Format(time.RFC3339), "get.error", err)

		return nil, err
	}

	if args != nil && args.Authorization != "" {
		req.Header.Set("Authorization", args.Authorization)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println(time.Now().Format(time.RFC3339), "get.error", err)

		return nil, err
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusOK:
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			fmt.Println(time.Now().Format(time.RFC3339), "get.error", err)

			return nil, err
		}

		result := &GetResult{
			Data:         string(body),
			ETag:         resp.Header.Get("ETag"),
			LastModified: resp.Header.Get("Last-Modified"),
		}

		return result, nil
	case http.StatusUnauthorized:
		return nil, fmt.Errorf("401 Unauthorized")
	case http.StatusNotFound:
		return nil, fmt.Errorf("404 Not Found")
	case http.StatusTooManyRequests:
		return nil, fmt.Errorf("429 Too Many Requests")
	}

	return nil, nil
}

var Message = struct {
	Unpack func(b []byte) *HttpStateMessageType
}{Unpack: func(b []byte) *HttpStateMessageType {
	var header int = int(b[0])

	if header == 0 {
		var length int = int(b[1])

		return &HttpStateMessageType{
			UUID:      string(b[2 : 2+length]),
			Timestamp: binary.BigEndian.Uint64(b[2+length : 2+length+8]),
			Type:      b[2+length+8],
			Value:     b[2+length+9:],
		}
	}

	return nil
}}

func Post(uuid string, data string, args *SetArgs) (int, error) {
	return Set(uuid, data, args)
}

func Put(uuid string, data string, args *SetArgs) (int, error) {
	return Set(uuid, data, args)
}

func Read(uuid string, args *GetArgs) (*GetResult, error) {
	return Get(uuid, args)
}

func Set(uuid string, data string, args *SetArgs) (int, error) {
	url := fmt.Sprintf("https://httpstate.com/%s", uuid)

	req, err := http.NewRequest("POST", url, bytes.NewBufferString(data))
	if err != nil {
		fmt.Println(time.Now().Format(time.RFC3339), "set.error", err)

		return 0, err
	}

	req.Header.Set("Content-Type", "text/plain;charset=UTF-8")

	if args != nil && args.Authorization != "" {
		req.Header.Set("Authorization", args.Authorization)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println(time.Now().Format(time.RFC3339), "set.error", err)

		return 0, err
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusUnauthorized:
		return 0, fmt.Errorf("401 Unauthorized")
	case http.StatusNotFound:
		return 0, fmt.Errorf("404 Not Found")
	case http.StatusRequestEntityTooLarge:
		return 0, fmt.Errorf("413 Content Too Large")
	}

	return resp.StatusCode, nil
}

func Write(uuid string, data string, args *SetArgs) (int, error) {
	return Set(uuid, data, args)
}

// HTTPState
type HttpState struct {
	Data *string
	ET   map[string][]HttpStateCallback
	UUID string
	WS   *websocket.Conn
}

type HttpStateCallback func(data *string)

type HttpStateMessageType struct {
	UUID      string
	Timestamp uint64
	Type      uint8
	Value     []byte
}

func New(uuid string) *HttpState {
	hs := &HttpState{
		Data: nil,
		ET:   make(map[string][]HttpStateCallback),
		UUID: uuid,
	}

	go hs.ws()

	return hs
}

func (hs *HttpState) Emit(_type string, data *string) *HttpState {
	if callbacks, ok := hs.ET[_type]; ok {
		for _, callback := range callbacks {
			callback(data)
		}
	}

	return hs
}

func (hs *HttpState) Get() *GetResult {
	result, err := Get(hs.UUID, nil)

	if err != nil || result == nil {
		return nil
	}

	hs.Data = &result.Data

	return result
}

func (hs *HttpState) Off(_type string, _callback HttpStateCallback) *HttpState {
	if callbacks, ok := hs.ET[_type]; ok {
		for i, callback := range callbacks {
			if fmt.Sprintf("%p", callback) == fmt.Sprintf("%p", _callback) {
				hs.ET[_type] = append(callbacks[:i], callbacks[i+1:]...)

				break
			}
		}

		if len(hs.ET[_type]) == 0 {
			delete(hs.ET, _type)
		}
	}

	return hs
}

func (hs *HttpState) On(_type string, _callback HttpStateCallback) *HttpState {
	hs.ET[_type] = append(hs.ET[_type], _callback)

	return hs
}

func (hs *HttpState) Post(data string) *int {
	return hs.Set(data)
}

func (hs *HttpState) Put(data string) *int {
	return hs.Set(data)
}

func (hs *HttpState) Read() *GetResult {
	return hs.Get()
}

func (hs *HttpState) Set(data string) *int {
	statusCode, err := Set(hs.UUID, data, nil)

	if err != nil {
		return nil
	}

	return &statusCode
}

func (hs *HttpState) Write(data string) *int {
	return hs.Set(data)
}

func (hs *HttpState) ws() {
	url := fmt.Sprintf("wss://httpstate.com/%s", hs.UUID)

	c, _, err := websocket.DefaultDialer.Dial(url, nil)
	if err != nil {
		fmt.Println("err:", err)

		return
	}
	defer c.Close()

	hs.WS = c

	if err := hs.WS.WriteMessage(websocket.TextMessage, []byte(fmt.Sprintf(`{"open":"%s"}`, hs.UUID))); err != nil {
		fmt.Println("err:", err)

		return
	}

	ping := time.NewTicker(time.Second * 30) // 30 SECONDS
	defer ping.Stop()

	pingClose := make(chan struct{})

	go func() {
		for {
			select {
			case <-ping.C:
				if err := c.WriteMessage(websocket.PingMessage, nil); err != nil {
					fmt.Println("err:", err)

					close(pingClose)

					return
				}
			case <-pingClose:
				return
			}
		}
	}()

	for {
		_, _data, err := c.ReadMessage()
		if err != nil {
			fmt.Println("err:", err)

			close(pingClose)

			return
		}

		var data *HttpStateMessageType = Message.Unpack(_data)

		if data != nil && data.UUID == hs.UUID && data.Type == 1 {
			var s string = string(data.Value)
			hs.Data = &s

			hs.Emit("change", hs.Data)
		}
	}
}

File: ruby/lib/httpstate.rb
# HTTP State, https://httpstate.com/
# Copyright (C) Alex Morales, 2026
#
# Unless otherwise stated in particular files or directories, this software is free software.
# You can redistribute it and/or modify it under the terms of the GNU Affero
# General Public License as published by the Free Software Foundation, either
# version 3 of the License, or (at your option) any later version.

# frozen_string_literal: true

require 'eventmachine'
require 'faye/websocket'
require 'json'
require 'net/http'
require 'socket'
require 'uri'

class HttpState
  VERSION = '0.1.1'

  attr_reader :data, :uuid, :ws

  def self.get(uuid)
    response = Net::HTTP.get(URI.join('https://httpstate.com', uuid))

    response
  rescue StandardError => e
    puts 'Error: ' + e.message

    nil
  end

  class Message
    class MessageType
      attr_accessor :uuid, :timestamp, :type, :value

      def initialize(uuid, timestamp, type, value)
        @uuid = uuid
        @timestamp = timestamp
        @type = type
        @value = value
      end
    end

    def self.unpack(b)
      length = b.getbyte(0)

      MessageType.new(
        b.byteslice(1, length).force_encoding("UTF-8"),
        b.byteslice(1+length, 8).unpack1("Q>"),
        b.getbyte(1+length+8),
        b.byteslice(1+length+9, b.bytesize-(1+length+9))
      )
    end
  end

  def self.message
    Message
  end

  def self.set(uuid, data)
    uri = URI.join('https://httpstate.com', uuid)

    request = Net::HTTP::Post.new(uri)
    request.body = data
    request.content_type = 'text/plain;charset=UTF-8'

    response = Net::HTTP.start(uri.hostname, uri.port, use_ssl:uri.scheme == 'https') do |http|
      http.request(request)
    end

    response.code
  rescue StandardError => e
    puts 'Error: ' + e.message

    nil
  end

  class << self
    alias read get
    alias post set
    alias put set
    alias write set
  end

  def initialize(uuid)
    @data = nil
    @et = nil
    @uuid = uuid
    @ws = nil

    Thread.new do
      EM.run do
        @ws = Faye::WebSocket::Client.new('wss://httpstate.com/' + @uuid)

        @ws.on :close do |_|
          puts '@ws.close'
        end

        @ws.on :error do |_|
          puts '@ws.error'
        end

        @ws.on :message do |event|
          data = self.class.message.unpack(event.data)

          if data &&
             data.uuid == @uuid &&
             data.type == 1
            @data = data.value

            emit('change', @data)
          end
        end

        @ws.on :open do |_|
          puts '@ws.open'

          @ws.send({ open:@uuid }.to_json)
        end
      end

      get()
    end
  end

  def emit(type, data = nil)
    if @et && @et[type]
      for callback in @et[type]
        if data.nil?
          callback.call
        else
          callback.call(data)
        end
      end
    end

    self
  end

  def get
    if @uuid
      data = self.class.get(@uuid)

      if(data != @data)
        # emit change
      end
      
      @data = data

      return @data
    end
  end

  def off(type, &callback)
    if @et && @et[type]
      if callback
        @et[type].delete(callback)
      end
      
      if !callback || @et[type].empty?
        @et.delete(type)
      end
    end

    self
  end

  def on(type, &callback)
    @et ||= {}
    @et[type] ||= []
    @et[type] << callback

    self
  end

  def set(data)
    self.class.set(@uuid, data) if @uuid
  end

  alias read get
  alias post set
  alias put set
  alias write set
end
